test/e2e/syncer/syncer_test.go

```diff
@@ -0,0 +1,126 @@
+	"github.com/stretchr/testify/require"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	apiextensionsclientset "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	kubernetesclientset "k8s.io/client-go/kubernetes"
+
+	"github.com/kcp-dev/kcp/pkg/syncer"
+	kubefixtures "github.com/kcp-dev/kcp/test/e2e/fixtures/kube"
+	"github.com/kcp-dev/kcp/test/e2e/framework"
+)
+
+//go:embed *.yaml
+var embeddedResources embed.FS
+
+func TestSyncerLifecycle(t *testing.T) {
+	t.Parallel()
+
+	upstreamServer := framework.SharedKcpServer(t)
+
+	t.Log("Creating an organization")
+	orgClusterName := framework.NewOrganizationFixture(t, upstreamServer)
+
+	t.Log("Creating a workspace")
+	wsClusterName := framework.NewWorkspaceFixture(t, upstreamServer, orgClusterName, "Universal")
+
+	resourcesToSync := sets.NewString("deployments.apps")
+	syncerFixture := framework.NewSyncerFixture(t, resourcesToSync, upstreamServer, orgClusterName, wsClusterName)
+
+	downstreamServer := syncerFixture.RunningServer
+	downstreamConfig := downstreamServer.DefaultConfig(t)
+	downstreamCrdClient, err := apiextensionsclientset.NewForConfig(downstreamConfig)
+	require.NoError(t, err)
+
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	t.Cleanup(cancelFunc)
+
+	t.Log("Install deployment crd in the downstream server to trigger the api importer")
+	kubefixtures.Create(t, downstreamCrdClient.ApiextensionsV1().CustomResourceDefinitions(),
+		metav1.GroupResource{Group: "apps.k8s.io", Resource: "deployments"},
+	)
+
+	syncerFixture.Start(t, ctx)
+
+	upstreamConfig := upstreamServer.DefaultConfig(t)
+	upstreamKubeClusterClient, err := kubernetesclientset.NewClusterForConfig(upstreamConfig)
+	require.NoError(t, err)
+	upstreamKubeClient := upstreamKubeClusterClient.Cluster(wsClusterName)
+
+	t.Log("Creating upstream namespace...")
+	upstreamNamespace, err := upstreamKubeClient.CoreV1().Namespaces().Create(ctx, &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "test-syncer",
+		},
+	}, metav1.CreateOptions{})
+	require.NoError(t, err)
+
+	t.Log("Creating upstream deployment...")
+
+	deploymentYAML, err := embeddedResources.ReadFile("deployment.yaml")
+	require.NoError(t, err, "failed to read embedded deployment")
+
+	var deployment *appsv1.Deployment
+	err = yaml.Unmarshal(deploymentYAML, &deployment)
+	require.NoError(t, err, "failed to unmarshal deployment")
+
+	// Since this test created a new workspace that initially lacked support for
+	// deployments, successful creation of a deployment in the workspace proves that
+	// the api importer imported the downstream type to the upstream server.
+	upstreamDeployment, err := upstreamKubeClient.AppsV1().Deployments(upstreamNamespace.Name).Create(ctx, deployment, metav1.CreateOptions{})
```

Comment:
Any chance this could fail due to timing - does this need to be in a poll/eventually block?

Reply:
The syncer waits for the gvrs to become available in the logical cluster before heartbeating, and the fixture Start method is checking for a ready status that depends on the heartbeat.

Reply:
I'll update the comment to be clear about this.

Reply:
Updated, PTAL